ateccx08/ateccx08.go

```diff
@@ -0,0 +1,258 @@
+func (d *Device) Wakeup() error {
+	d.bus.Tx(uint16(0x0), []byte{0x00}, nil)
+	time.Sleep(1500 * time.Microsecond)
+	d.bus.Tx(uint16(d.Address), []byte{0x00}, nil)
+	time.Sleep(maxCommandTime)
+
+	var status [4]byte
+	if err := d.readResponse(status[:]); err != nil {
+		return err
+	}
+
+	if status[0] != StatusAfterWake {
+		return ErrWakeup
+	}
+
+	return nil
+}
+
+// Sleep puts the ATECC to sleep.
+func (d *Device) Sleep() {
+	d.bus.Tx(uint16(d.Address), []byte{0x01}, nil)
+	time.Sleep(time.Millisecond)
+}
+
+// Idle puts the ATECC in idle mode.
+func (d *Device) Idle() {
+	d.bus.Tx(uint16(d.Address), []byte{0x02}, nil)
+	time.Sleep(time.Millisecond)
+}
+
+type ATECCVersion uint16
+
+func (at ATECCVersion) String() string {
+	switch at {
+	case ATECC508:
+		return "ATECC508"
+	case ATECC608:
+		return "ATECC608"
+	case ATECCNone:
+		return "No ATECCx08"
+	default:
+		return "Unknown"
+	}
+}
+
+// Version returns what version of ATECC is being used.
+// Either ATECC508, ATECC608, or ATECCNone.
+func (d *Device) Version() (ATECCVersion, error) {
+	var version [4]byte
+	d.Wakeup()
+	defer d.Idle()
+
+	d.sendCommand(cmdInfo, 0x00, 0, nil)
+
+	time.Sleep(maxCommandTime)
+	if err := d.readResponse(version[:]); err != nil {
+		return ATECCNone, err
+	}
+
+	return ATECCVersion(uint16(version[2])<<8 | uint16(version[3])&0xf000), nil
+}
+
+// Random returns an array of 32 byte-sized random numbers.
+func (d *Device) Random() ([32]byte, error) {
+	var random [32]byte
+	d.Wakeup()
+	defer d.Idle()
+
+	d.sendCommand(cmdRandom, 0x00, 0, nil)
+	time.Sleep(23 * time.Millisecond)
+
+	err := d.readResponse(random[:])
+	return random, err
+}
+
+// Read reads from the device memory.
+func (d *Device) Read(zone, address int, data []byte) error {
+	d.Wakeup()
+	defer d.Idle()
+
+	d.sendCommand(cmdRead, byte(zone), uint16(address), nil)
+	time.Sleep(5 * time.Millisecond)
+
+	return d.readResponse(data)
+}
+
+// IsLocked checks to see if the ATECC is locked.
+// Config zone (0) must be locked to generate random numbers.
+func (d *Device) IsLocked() bool {
+	return d.IsZoneLocked(0)
+}
+
+// IsZoneLocked checks to see if a specific zone in the ATECC is locked.
+func (d *Device) IsZoneLocked(zone int) bool {
+	var config [4]byte
+
+	if zone < 0 || zone > 8 {
+		return false
+	}
+
+	switch zone {
+	case 0, 1:
+		if err := d.Read(0, 0x15, config[:]); err != nil {
+			return false
+		}
+
+		// LockConfig
+		loc := 3
+
+		// LockData
+		if zone == 1 {
+			loc = 2
+		}
+
+		if config[loc] == 0 {
+			return true
+		}
+
+	default:
+		if err := d.Read(0, 0x16, config[:]); err != nil {
+			return false
+		}
+
+		slot := byte(zone<<2) | 2
+
+		if (config[0] & slot) == 0 {
+			return true
+		}
+
+		return false
+	}
+
+	return false
+}
+
+// Lock locks a zone in the device.
+// Note that you cannot unlock a device zone once locked,
+// so make sure you know what you are doing!
+func (d *Device) Lock(zone int) error {
```

Comment:
Any link where to read about zone locking and what can go wrong? Covered in datasheet, perhaps?

Reply:
> For the ATECC608A-TNGTLS, the Configuration zone has already been locked and the access policies of the Data zone have already been enforced. However, several of the data slots can still be updated through the use of other commands. If so desired, some of these slots can be permanently locked from future updates by using the Slot Locking mode of the Lock command.
> The Slot Lock mode of the Lock command is used to individually lock slots. Any slot that has the Lockable bit set and has not previously been slot-locked can be locked to prevent any further updates. This process is not reversible. Once a slot has been locked, it is permanently locked. In this mode of operation, the Summary CRC and Data fields are ignored.